translator/trace/protospan_translation.go

```diff
@@ -102,25 +136,171 @@ func AttributeValueToString(attr pdata.AttributeValue, jsonLike bool) string {
 	}
+}
+
+// AttributeMapToMap converts an OTLP AttributeMap to a standard go map
+func AttributeMapToMap(attrMap pdata.AttributeMap) map[string]interface{} {
+	rawMap := make(map[string]interface{})
+	attrMap.ForEach(func(k string, v pdata.AttributeValue) {
+		switch v.Type() {
+		case pdata.AttributeValueSTRING:
+			rawMap[k] = v.StringVal()
+		case pdata.AttributeValueINT:
+			rawMap[k] = v.IntVal()
+		case pdata.AttributeValueDOUBLE:
+			rawMap[k] = v.DoubleVal()
+		case pdata.AttributeValueBOOL:
+			rawMap[k] = v.BoolVal()
+		case pdata.AttributeValueNULL:
+			rawMap[k] = nil
+		case pdata.AttributeValueMAP:
+			rawMap[k] = AttributeMapToMap(v.MapVal())
+		case pdata.AttributeValueARRAY:
+			rawMap[k] = AttributeArrayToSlice(v.ArrayVal())
+		}
+	})
+	return rawMap
+}
+
+func AttributeArrayToSlice(attrArray pdata.AnyValueArray) []interface{} {
+	rawSlice := make([]interface{}, 0, attrArray.Len())
+	for i := 0; i < attrArray.Len(); i++ {
+		v := attrArray.At(i)
+		switch v.Type() {
+		case pdata.AttributeValueSTRING:
+			rawSlice = append(rawSlice, v.StringVal())
+		case pdata.AttributeValueINT:
+			rawSlice = append(rawSlice, v.IntVal())
+		case pdata.AttributeValueDOUBLE:
+			rawSlice = append(rawSlice, v.DoubleVal())
+		case pdata.AttributeValueBOOL:
+			rawSlice = append(rawSlice, v.BoolVal())
+		case pdata.AttributeValueNULL:
+			rawSlice = append(rawSlice, nil)
+		default:
+			rawSlice = append(rawSlice, "<Invalid array value>")
+		}
+	}
+	return rawSlice
+}
 
-	// TODO: Add support for ARRAY type.
+// UpsertStringToAttributeMap upserts a string value to the specified key as it's native OTLP type
+func UpsertStringToAttributeMap(key string, val string, dest pdata.AttributeMap, omitSimpleTypes bool) {
+	switch DetermineValueType(val, omitSimpleTypes) {
+	case pdata.AttributeValueINT:
+		iVal, _ := strconv.ParseInt(val, 10, 64)
+		dest.UpsertInt(key, iVal)
+	case pdata.AttributeValueDOUBLE:
+		fVal, _ := strconv.ParseFloat(val, 64)
+		dest.UpsertDouble(key, fVal)
+	case pdata.AttributeValueBOOL:
+		bVal, _ := strconv.ParseBool(val)
+		dest.UpsertBool(key, bVal)
+	case pdata.AttributeValueMAP:
+		var attrs map[string]interface{}
+		err := json.Unmarshal([]byte(val), &attrs)
+		if err == nil {
+			attrMap := pdata.NewAttributeValueMap()
+			jsonMapToAttributeMap(attrs, attrMap.MapVal())
+			dest.Upsert(key, attrMap)
+		} else {
+			dest.UpsertString(key, "")
+		}
+	case pdata.AttributeValueARRAY:
+		var jArray []interface{}
+		err := json.Unmarshal([]byte(val), &jArray)
+		if err == nil {
+			attrArr := pdata.NewAttributeValueArray()
+			jsonArrayToAttributeArray(jArray, attrArr.ArrayVal())
+			dest.Upsert(key, attrArr)
+		} else {
+			dest.UpsertString(key, "")
+		}
+	default:
+		dest.UpsertString(key, val)
+	}
+}
+
+// DetermineValueType returns the native OTLP attribute type the string translates to.
+func DetermineValueType(value string, omitSimpleTypes bool) pdata.AttributeValueType {
```

Comment:
I am not sure I understand why this smartness exists in our translation. I was not aware we had this for Zipkin, and I don't understand why we extend this to other translations either.
Do you know why we had this for Zipkin and why we would want it for other formats? Jaeger has proper support for all data types we need, I don't know why we would need a smart conversion like this.